include/v4l2_camera/rate_bound_status.hpp

```diff
@@ -0,0 +1,281 @@
+ */
+class RateBoundStatus : public diagnostic_updater::DiagnosticTask
+{
+private:
+  // Helper struct to express state machine nodes
+  struct StateBase
+  {
+    StateBase(const unsigned char lv, const std::string m)
+        : level(lv), num_observation(1), msg(m) {}
+
+    unsigned char level;
+    size_t num_observation;
+    std::string msg;
+  };
+
+  struct Stale : public StateBase
+  {
+    Stale()
+        : StateBase(diagnostic_msgs::msg::DiagnosticStatus::STALE,
+                    "Topic has not been received yet") {}
+  };
+
+  struct Ok : public StateBase
+  {
+    Ok()
+        : StateBase(diagnostic_msgs::msg::DiagnosticStatus::OK,
+                    "Rate is reasonable") {}
+  };
+
+  struct Warn : public StateBase
+  {
+    Warn()
+        : StateBase(diagnostic_msgs::msg::DiagnosticStatus::WARN,
+                    "Rate is within warning range") {}
+  };
+
+  struct Error : public StateBase
+  {
+    Error()
+        : StateBase(diagnostic_msgs::msg::DiagnosticStatus::ERROR,
+                    "Rate is out of valid range") {}
+  };
+
+  using StateHolder = std::variant<Stale, Ok, Warn, Error>;
+
+ public:
+  /**
+   * \brief Constructs RateBoundstatus, which inherits diagnostic_updater::DiagnosticTask.
+   *
+   * \param ok_params The pair of min/max frequency for the topic rate to be recognized as "OK".
+   * \param warn_params The pair of min/max frequency for the topic rate to be recognized as "WARN".
+   * These values should have a wider range than `ok_params`.
+   * \param num_frame_transition The number of the successive observations for the status
+   * transition. E.g., the status will not be changed from OK to WARN until successive
+   * `num_frame_transition` WARNs are observed.
+   * \param immediate_error_report If true (default), errors related to the rate bounds will be
+   * reported immediately once observed; otherwise, the hysteresis damping method using
+   * `num_frame_transition` will be adopted
+   * \param name The arbitrary string to be assigned for this diagnostic task.
+   * This name will not be exposed in the actual published topics.
+   */
+  RateBoundStatus(const RateBoundStatusParam& ok_params,
+                  const RateBoundStatusParam& warn_params,
+                  const size_t num_frame_transition = 1,
+                  const bool immediate_error_report = true,
+                  std::string name = "rate bound check")
+      : DiagnosticTask(name), ok_params_(ok_params), warn_params_(warn_params),
+        num_frame_transition_(num_frame_transition),
+        immediate_error_report_(immediate_error_report), zero_seen_(false),
+        candidate_state_(Stale{}), current_state_(Stale{})
+  {
+    if (num_frame_transition < 1) {
+      num_frame_transition_ = 1;
+    }
+
+    // Confirm `warn_params` surely has wider range than `ok_params`
+    if (!(warn_params_.min_frequency < ok_params_.min_frequency &&
+          ok_params_.max_frequency< warn_params_.max_frequency)) {
```

Comment:
This could be simplified (de-morgan), Nebula Clang-Tidy flags this:

```suggestion
    if (warn_params_.min_frequency >= ok_params_.min_frequency ||
      ok_params_.max_frequency >= warn_params_.max_frequency) {
```